File: frontend/lib/api/studio.ts
import { http, qs } from "./http";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000/api/v1";
const HOST = API_URL.replace(/\/api\/v1\/?$/, "");

/** Resolve a backend-relative media path (e.g. /media/generated/x.png) to an absolute URL. */
export function mediaUrl(path: string): string {
  if (!path) return "";
  if (path.startsWith("http")) return path;
  return `${HOST}${path.startsWith("/") ? "" : "/"}${path}`;
}

export type AssetKind = "copy" | "image" | "video_plan";

export interface CopyVariant {
  text: string;
  hashtags: string[];
  angle: string;
  chars: number;
}

export interface CopyGroup {
  platform: string;
  variants: CopyVariant[];
}

export interface CopyResponse {
  asset_id: string | null;
  results: CopyGroup[];
  generated_by: string;
}

export type ImageQuality = "low" | "medium" | "high";

/** "off" | "auto" | a position — where to composite the workspace brand logo. */
export type LogoPlacement =
  | "off"
  | "auto"
  | "top-left"
  | "top-center"
  | "top-right"
  | "center-left"
  | "center"
  | "center-right"
  | "bottom-left"
  | "bottom-center"
  | "bottom-right";

export interface ImageChatTurn {
  role: "user" | "assistant";
  text: string;
}

export type ImageJobStatus = "queued" | "processing" | "succeeded" | "failed";

/** A background image generation. Created by POST, then polled. */
export interface ImageJob {
  id: string;
  status: ImageJobStatus;
  mode: "generate" | "chat";
  url: string | null;
  asset_id: string | null;
  reply: string | null;
  operation: string | null;
  used_base: string | null;
  logo_applied: string | null;
  logo_note: string | null;
  error: string | null;
}

export interface GeneratedAsset {
  id: string;
  kind: AssetKind;
  prompt: string;
  platform: string | null;
  title: string | null;
  payload: Record<string, unknown> | null;
  image_url: string | null;
  created_at: string;
}

export interface PromptResult {
  prompt: string;
  negative_prompt: string;
  style_tips: string[];
  notes: string;
}

export interface BrandKit {
  brand_logo_url: string | null;
  brand_colors: string[] | null;
}

export const studioApi = {
  copy: (body: { prompt: string; platforms: string[]; count?: number; tone?: string }) =>
    http.post<CopyResponse>("/studio/copy", body),
  buildPrompt: (body: { intent: "image" | "video"; brief: string }) =>
    http.post<PromptResult>("/studio/prompt", body),
  image: (body: {
    prompt: string;
    size?: string;
    quality?: ImageQuality;
    style?: string;
    draft?: boolean;
    input_image_url?: string;
    as_logo?: boolean;
    brand_logo?: LogoPlacement;
  }) => http.post<ImageJob>("/studio/image", body),
  imageChat: (body: {
    instruction: string;
    attachment_url?: string;
    previous_image_url?: string;
    history?: ImageChatTurn[];
    brand_logo?: LogoPlacement;
  }) => http.post<ImageJob>("/studio/image/chat", body),
  imageJob: (id: string) => http.get<ImageJob>(`/studio/image/jobs/${id}`),
  assets: (kind?: AssetKind) => http.get<{ items: GeneratedAsset[] }>(`/studio/assets${qs({ kind })}`),
  deleteAsset: (id: string) => http.del<void>(`/studio/assets/${id}`),
  brand: () => http.get<BrandKit>("/studio/brand"),
  setBrand: (logo_url: string, colors: string[]) =>
    http.patch<BrandKit>("/studio/brand", { logo_url, colors }),
  clearBrand: () => http.del<BrandKit>("/studio/brand"),
  saveCaption: (body: {
    title: string;
    platform: string;
    text: string;
    hashtags: string[];
    angle?: string;
    brief?: string;
  }) => http.post<GeneratedAsset>("/studio/captions", body),
};

File: backend/app/schemas/studio.py
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class AssetKind(str, Enum):
    COPY = "copy"
    IMAGE = "image"
    VIDEO_PLAN = "video_plan"


class CopyRequest(BaseModel):
    prompt: str = Field(..., min_length=3, max_length=4000)
    platforms: List[str] = Field(default_factory=lambda: ["instagram"], min_length=1, max_length=6)
    count: int = Field(2, ge=1, le=3, description="Variants per channel")
    tone: str = "confident, friendly"
    # A fresh generation is a draft, not a save — persisting a caption is a
    # deliberate, titled action now (see SaveCaptionRequest / POST /captions).
    save: bool = False


class SaveCaptionRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=120)
    platform: str = Field(..., min_length=1, max_length=40)
    text: str = Field(..., min_length=1, max_length=4000)
    hashtags: List[str] = Field(default_factory=list, max_length=20)
    angle: str = Field("", max_length=120)
    brief: str = Field("", max_length=4000)


class ImageQuality(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class ImageRequest(BaseModel):
    prompt: str = Field(..., min_length=3, max_length=6000)
    size: str = "1024x1024"
    quality: ImageQuality = ImageQuality.MEDIUM
    style: str = ""
    draft: bool = Field(False, description="Use the cheaper gpt-image-1-mini model")
    input_image_url: Optional[str] = Field(None, description="An uploaded image to edit / use as a base")
    as_logo: bool = False
    save: bool = True
    brand_logo: Optional[str] = Field(
        None,
        description=(
            "Composite the workspace brand logo onto the result. "
            "'off'/None = never; 'auto' = only if the prompt asks for a logo, "
            "at the spot it names; or an explicit position "
            "(top-left, top-right, bottom-left, bottom-right, center, …)."
        ),
    )


class PromptRequest(BaseModel):
    intent: str = Field("image", pattern="^(image|video)$")
    brief: str = Field(..., min_length=3, max_length=3000)


class PromptResponse(BaseModel):
    prompt: str
    negative_prompt: str = ""
    style_tips: list[str] = []
    notes: str = ""


class CopyVariant(BaseModel):
    text: str
    hashtags: List[str] = []
    angle: str = ""
    chars: int = 0


class CopyGroup(BaseModel):
    platform: str
    variants: List[CopyVariant] = []


class CopyResponse(BaseModel):
    asset_id: Optional[str] = None
    results: List[CopyGroup]
    generated_by: str


class ImageResponse(BaseModel):
    asset_id: Optional[str] = None
    id: str
    url: str
    prompt: str
    size: str
    quality: str = "medium"
    model: str = ""
    style: str = ""
    logo_applied: Optional[str] = Field(None, description="position the brand logo was placed, if any")
    logo_note: Optional[str] = Field(None, description="why the logo wasn't placed, if it was requested")


class ChatTurn(BaseModel):
    role: str = "user"
    text: str = ""


class ImageChatRequest(BaseModel):
    instruction: str = Field(..., min_length=1, max_length=6000)
    attachment_url: Optional[str] = Field(None, description="A photo the user just attached")
    previous_image_url: Optional[str] = Field(None, description="The last image the tool made in this chat")
    history: List[ChatTurn] = Field(default_factory=list, max_length=40)
    brand_logo: Optional[str] = Field(
        None,
        description="off/None, 'auto', or an explicit position — same as ImageRequest.brand_logo",
    )


class ImageChatResponse(BaseModel):
    asset_id: Optional[str] = None
    id: str
    url: str
    prompt: str
    size: str
    quality: str = "medium"
    model: str = ""
    reply: str = ""
    operation: str = "generate"
    used_base: str = "none"
    logo_applied: Optional[str] = None
    logo_note: Optional[str] = None


class AssetResponse(BaseModel):
    id: str
    kind: AssetKind
    prompt: str
    platform: Optional[str] = None
    title: Optional[str] = None
    payload: Optional[Dict[str, Any]] = None
    image_url: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AssetList(BaseModel):
    items: List[AssetResponse]


class ImageJobResponse(BaseModel):
    """Poll target for a background image generation."""
    id: str
    status: str  # queued | processing | succeeded | failed
    mode: str    # generate | chat
    # populated once status == succeeded
    url: Optional[str] = None
    asset_id: Optional[str] = None
    reply: Optional[str] = None
    operation: Optional[str] = None
    used_base: Optional[str] = None
    logo_applied: Optional[str] = None
    logo_note: Optional[str] = None
    error: Optional[str] = None

    model_config = {"from_attributes": True}


class BrandKitResponse(BaseModel):
    brand_logo_url: Optional[str] = None
    brand_colors: Optional[List[str]] = None


class BrandKitUpdate(BaseModel):
    logo_url: str = ""
    colors: List[str] = Field(default_factory=list, max_length=6)
